nomenklatura/store/sql.py

```diff
@@ -0,0 +1,167 @@
+                    prop_type=istmt.excluded.prop_type,
+                    target=istmt.excluded.target,
+                    lang=istmt.excluded.lang,
+                    original_value=istmt.excluded.original_value,
+                    last_seen=istmt.excluded.last_seen,
+                ),
+            )
+            with self.store.engine.connect() as conn:
+                conn.begin()
+                conn.execute(stmt)
+                conn.commit()
+        self.batch = set()
+        self.batch_size = 0
+
+    def add_statement(self, stmt: Statement) -> None:
+        if self.batch is None:
+            self.batch = set()
+        if stmt.entity_id is None:
+            return
+        if self.batch_size >= self.BATCH_STATEMENTS:
+            self.flush()
+        canonical_id = self.store.resolver.get_canonical(stmt.entity_id)
+        stmt.canonical_id = canonical_id
+        self.batch.add(stmt)
+        self.batch_size += 1
+
+    def pop(self, entity_id: str) -> List[Statement]:
+        self.flush()
+        table = self.store.table
+        q = select(table).where(table.c.entity_id == entity_id)
+        q_delete = delete(table).where(table.c.entity_id == entity_id)
+        statements: List[Statement] = list(self.store._iterate_stmts(q))
+        with self.store.engine.connect() as conn:
+            conn.begin()
+            conn.execute(q_delete)
+            conn.commit()
+        return statements
+
+
+class SqlView(View[DS, CE]):
+    def __init__(
+        self, store: SqlStore[DS, CE], scope: DS, external: bool = False
+    ) -> None:
+        super().__init__(store, scope, external=external)
+        self.store: SqlStore[DS, CE] = store
+
+    def get_entity(self, id: str) -> Optional[CE]:
+        table = self.store.table
+        ids = [str(i) for i in self.store.resolver.connected(id)]
+        q = select(table).where(
+            table.c.entity_id.in_(ids), table.c.dataset.in_(self.dataset_names)
+        )
+        for proxy in self.store._iterate(q):
+            return proxy
+        return None
+
+    def get_inverted(self, id: str) -> Generator[Tuple[Property, CE], None, None]:
+        table = self.store.table
+        q = (
+            select(table)
+            .where(table.c.prop_type == "entity", table.c.value == id)
+            .distinct(table.c.value)
+        )
+        for stmt in self.store._iterate_stmts(q):
+            if stmt.canonical_id is not None:
+                entity = self.get_entity(stmt.canonical_id)
+                if entity is not None:
+                    for prop, value in entity.itervalues():
+                        if value == id and prop.reverse is not None:
+                            yield prop.reverse, entity
+
+    def entities(self) -> Generator[CE, None, None]:
+        table: Table = self.store.table
+        q = (
+            select(table)
+            .where(table.c.dataset.in_(self.dataset_names))
+            .order_by("entity_id")
```

Comment:
Does this need to be `canonical_id`? Otherwise it'll return fragmented entities, no?

Reply:
totally right. fixed here: https://github.com/opensanctions/nomenklatura/pull/125/commits/c1dc320ca70e8a119df8d34b388eab1c8e13e19d

but, `canonical_id` column was nullable in the sql table, which should not, right?